crates/ntx-builder/src/state/mod.rs

```diff
@@ -148,36 +178,60 @@ impl State {
                 }
             },
         }
+
+        Ok(())
     }
 
     /// Handles a [`MempoolEvent::TransactionAdded`] event.
     ///
     /// Note that this will include our own network transactions as well as user submitted
     /// transactions.
-    fn add_transaction(
+    async fn add_transaction(
         &mut self,
         id: TransactionId,
         nullifiers: Vec<Nullifier>,
         network_notes: Vec<NetworkNote>,
         account_delta: Option<AccountUpdateDetails>,
-    ) {
+    ) -> anyhow::Result<()> {
         // Skip transactions we already know about.
         //
         // This can occur since both ntx builder and the mempool might inform us of the same
         // transaction. Once when it was submitted to the mempool, and once by the mempool event.
         if self.inflight_txs.contains_key(&id) {
-            return;
+            return Ok(());
         }
 
         let mut tx_impact = TransactionImpact::default();
         if let Some(update) = account_delta.and_then(NetworkAccountUpdate::from_protocol) {
-            tx_impact.account_delta = Some(update.prefix());
-            self.account_or_default(update.prefix()).add_delta(update);
+            let prefix = update.prefix();
+            match update {
+                NetworkAccountUpdate::New(account) => {
+                    let account_state = AccountState::uncommitted(account);
+                    self.accounts.insert(prefix, account_state);
+                    self.queue.push_back(prefix);
+                },
+                NetworkAccountUpdate::Delta(account_delta) => {
+                    self.account_or_load(prefix)
+                        .await
+                        .context("failed to load account")?
+                        .context("account with delta not found")?
+                        .add_delta(&account_delta);
+                },
+            }
+
+            tx_impact.account_delta = Some(prefix);
         }
         for note in network_notes {
             tx_impact.notes.insert(note.nullifier());
             self.nullifier_idx.insert(note.nullifier(), note.account_prefix());
-            self.account_or_default(note.account_prefix()).add_note(note);
+            // Skip notes which target a non-existent network account.
+            if let Some(account) = self
+                .account_or_load(note.account_prefix())
+                .await
+                .context("failed to load account")?
+            {
+                account.add_note(note);
+            }
```

Comment:
This could be an issue if many notes get created for accounts that don't exist. We probably need a blocklist which prevents re-querying the store for the same non-existent account.